### DISCUSSION TITLE
[GITLAB-CI-RUNNER] How to install private NPM packages

### DISCUSSION THREAD

**Discussion by felixranesberger** [renovatebot/renovate#19707] [Request Help] [answered]
Hi, I would like to install a private NPM package from my Gitlab Instance using the Renovate Gitlab CI runner. 

When running a `npm install` in my Gitlab CI build stage I have to set a valid token first, to allow an installation of the package. 

`npm config set -- '//gitlab.domain.de/api/v4/projects/321/packages/npm/:_authToken' "${CI_JOB_TOKEN}"`

I've seen some [examples](https://docs.renovatebot.com/getting-started/private-packages/#add-hostrule-to-bots-config) using module.exports and process (environment) variables, but I'm not sure, in what combination these solutions are used.

How can I configure the CI runner to use the `CI_JOB_TOKEN` of the pipeline?

**Comment by rarkins** [comment, marked as answer]
You might want to try `detectGlobalManagerConfig=true` in config to tell Renovate to read from the global `.npmrc` file which is set by `npm config`

**Reply by felixranesberger** [reply]
I've seen that this option only works for private NPM packages. 

Is there also a good solution for installing private composer packages using the Gitlab package registry?

**Reply by felixranesberger** [reply]
Currently I'm trying this for installing the NPM package in my `config.js`, but without success.

```js 
module.exports = {
  hostRules: [
    {
      matchHost: 'gitlab.example.com',
      hostType: 'npm',
      token: process.env.CI_JOB_TOKEN,
    },
  ],
};
```

**Reply by rarkins** [reply]
It need to match the full url/path as in your npmrc

**Reply by felixranesberger** [reply]
Hmh, I can't get it to work using this configuration:

### Error message:
`Failed to look up npm dependency @organisation-team/package-name`

### Configuration:
#### renovate.json
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "packageRules": [
    {
      "groupName": "TYPO3",
      "matchPackagePrefixes": [
        "typo3"
      ],
      "allowedVersions": "<12.0.0"
    }
  ]
}
```

#### config.js
```js
module.exports = {
  hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://gitlab.domain.com/api/v4/packages/npm/',
      token: process.env.CI_JOB_TOKEN,
    },
  ],
  npmrc: '@organisation-team:registry=https://gitlab.domain.de/api/v4/packages/npm/'
};
```

### .gitlab-ci.yml
```yaml
include:
  - project: 'organisation-Team/Tools/renovate-runner'
    file: '/templates/renovate-dind.gitlab-ci.yml'

renovate:
  script:
    - renovate $RENOVATE_EXTRA_FLAGS
```

#### $RENOVATE_EXTRA_FLAGS
`--git-author=bot@renovate --autodiscover=true --onboarding=true --autodiscover-filter=organisation-Team/customer/*`

**Reply by felixranesberger** [reply]
There seems to be an issue resolving the `@organisation-team/package-name` to the correct Gitlab URL. Shouldn't this be resolved by the `npmrc` part in the `config.js` file?

```json
{
  "name": "renovate",
  "hostname": "runner-vs97sfq-project-602-concurrent-0",
  "pid": 15,
  "level": 20,
  "logContext": "xa9HBKwKYSOlhz25Rk_q7",
  "repository": "organisation-team/customer-group/repository",
  "msg": "GET https://registry.npmjs.org/@organisation-team%2Fpackage = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=551)",
  "time": "2023-01-07T18:37:52.076Z",
  "v": 0
}
```

**Reply by rarkins** [reply]
I'm not sure why, but it looks like that `npmrc` setting is not being applied. What do you find if you search for the string `npmrc` within debug logs?

**Reply by felixranesberger** [reply]
Hi, sorry for the late response. 
Maybe I haven't understood the configuration process correctly. 

I have a Gitlab Renovate Runner project mirror, which I haven't changed at all. 
Then I have a Gitlab Renovate project, which executes a pipeline every 24 hours. 

The part where I might be wrong, is where I should put the configuration files. 

I have the following two configuration files, inside my Renovate project, where the scheduler is executed: 

**config.js**
```js
module.exports = {
  endpoint: 'https://gitlab.organisation.de/api/v4/',
  platform: 'gitlab',
  hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://gitlab.organisation.de/api/v4/packages/npm/',
      token: process.env.CI_JOB_TOKEN,
    },
  ],
}
```

**renovate.json:**
```json
{
  "npmrc": "@organisation-team:registry=https://gitlab.organisation.de/api/v4/packages/npm/"
}
```

In my Gitlab project, which Renovate should update the dependencies, I have a **renovate.json** configuration file as well. 

**renovate.json:**
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "packageRules": [
    {
      "groupName": "TYPO3",
      "matchPackagePrefixes": [
        "typo3"
      ],
      "allowedVersions": "<12.0.0"
    }
  ]
}
```

Can you point me to the error, I can't figure it out. 
Your help is much appreciated :)

**Reply by felixranesberger** [reply]
<img width="588" alt="Screenshot 2023-02-27 at 21 40 09" src="https://user-images.githubusercontent.com/52704891/221680558-706b9f63-5776-4985-8e84-bc9085df5a25.png">

**Reply by rarkins** [reply]
The first `npmrc` above (i.e. the middle file of 3) won't be read when processing "my Gitlab project, which Renovate should update the dependencies". You can do any of these:
- Add the `npmrc` config line to `config.js`, or
- Add the `npmrc` line to each project's `renovate.json`, or
- Commit that `npmrc` to each project (so it's part of the repo)
